### ISSUE TITLE
content-security-policy set by twitter.com causes injected scriptlets to not run

### ISSUE THREAD

**Issue by anvakl** [uBlockOrigin/uBlock-issues#235]
### Prerequisites

- [x] I verified that this is not a filter issue
    - Filter issues MUST be reported at [filter issue tracker](https://github.com/uBlockOrigin/uAssets/issues)
    
- [x] This is not a support issue or a question
    - Support issues and questions are handled at [/r/uBlockOrigin](https://old.reddit.com/r/uBlockOrigin/)
    
- [ ] I performed a cursory search of the issue tracker to avoid opening a duplicate issue
    - Your issue may already be reported.
- I tried to reproduce the issue when...
    - [x] uBlock Origin is the only extension
    - [ ] uBlock Origin with default lists/settings
    - [ ] using a new, unmodified browser profile
- [ ] I am running the latest version of uBlock Origin
- [ ] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I report is not a normal behavior

### Description

content security policy set by twitter doesnot allow inline scripts without` nonce=` attribute to run.

csp set by twitter is

`“script-src https://connect.facebook.net https://cm.g.doubleclick.net https://ssl.google-analytics.com https://graph.facebook.com https://twitter.com 'unsafe-eval' https://*.twimg.com https://api.twitter.com https://analytics.twitter.com https://publish.twitter.com https://ton.twitter.com https://syndication.twitter.com https://www.google.com https://t.tellapart.com https://platform.twitter.com https://www.google-analytics.com blob: 'nonce-P+pysraZ1bM1TVgufAjzPw==' https://twitter.com”`

I got that from beowser console. 

### A specific URL where the issue occurs

https://twitter.com/gorhill

### Steps to Reproduce

1. Create a rule `twitter.com##+js(addEventListener-logger.js)`
2. open https://twitter.com/gorhill
3. open browser console to see log.

### Expected behavior:

Scriptlet logs all event listeners

### Actual behavior:

Get an error saying : Content Security Policy: The page’s settings blocked the loading of a resource at self

Palemoon displays full csp.
Firefox says it is because of script-src

### Your environment

Firefox 60 ESR with uBlock1.16.21rc1
Palemoon 28.1 with uBlock-legacy1.16.4.4

Windows 8.1 x64

**Comment by uBlock-user** [comment]
Browser bug - 

https://bugzilla.mozilla.org/show_bug.cgi?id=1267027
https://bugzilla.mozilla.org/show_bug.cgi?id=866522

**Comment by joey04** [comment]
Question:
Would a workaround be to use a regular content script, like uBO-extra or uBO's subscriber link finder?

**Comment by uBlock-user** [comment]
Firefox does have a [contentScripts](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/contentScripts) API which could be used if it's possible.

**Comment by gwarser** [comment]
Duplicate of https://github.com/gorhill/uBlock/issues/2823

**Comment by gorhill** [comment]
> Duplicate of https://github.com/gorhill/uBlock/issues/2823

Different, that other one was about redirect to `data:` URI causing the failure -- and this was fixed by using web-accessible resources. The one here is about inline `script` tags injected in the DOM.

**Comment by gwarser** [comment]
Then "see also".

**Comment by JustOff** [comment]
Since Firefox [bug #1267027](https://bugzilla.mozilla.org/show_bug.cgi?id=1267027) is still not fixed, and @gorhill is clearly [against uBO lowering CSP rules](https://github.com/gorhill/uBlock/issues/2823#issuecomment-317943668) set by sites, I created a separate add-on called [Scriptlet Doctor](https://github.com/JustOff/scriptlet-doctor), which provides such a workaround for both Firefox current and legacy-based browsers.

By default, Scriptlet Doctor modifies CSP only for a specific list of domains that can be configured. Currently, this list is pre-filled with domains requested by [RU AdList](https://forums.lanik.us/viewtopic.php?f=102&t=22512) admin @dimisa-RUAdList.

Hope someone finds it useful.

**Comment by evilpie** [comment]
There is a relatively simple workaround and I am not sure if you have considered doing that @gorhill. The problem is that uBlock uses an actual inline script for injection, but Firefox only bypasses the CSP for `<script>` loads triggered via an URL.

The following change makes uBlock immune to the page's CSP:
```patch
+++ b/src/js/contentscript.js
@@ -468,8 +468,11 @@ vAPI.injectScriptlet = function(doc, text) {
     let script;
     try {
         script = doc.createElement('script');
-        script.appendChild(doc.createTextNode(text));
+        let blob = new Blob([decodeURIComponent(scriptlets)], {type: 'text/javascript'});
+        let url = document.URL.createObjectURL(blob);
+        script.url = url;
         (doc.head || doc.documentElement || doc).appendChild(script);
+        document.URL.revokeObjectURL(url);
     } catch (ex) {
     }
     if ( script ) {
```

**Comment by evilpie** [comment]
> Immune to blob: being blocked by page's CSP ?

Sure this works even with `Content-Security-Policy: script-src 'none';`

**Comment by gorhill** [comment]
@evilpie Thanks for the suggested fix, seems to work fine in both Firefox and Chromium, will publish a new version with fix as suggested.

**Comment by krystian3w** [comment]
#2404

**Comment by krystian3w** [comment]
https://github.com/AdguardTeam/AdguardBrowserExtension/issues/1733

**Comment by gorhill** [comment]
The mitigation for this in uBO is to try to insert the script synchronously first, and if this fails, use the `blob:` approach:
https://github.com/gorhill/uBlock/blob/7dba69dc48d2114648ca237e4b69031d5db8626b/platform/firefox/vapi-background-ext.js#L364-L387
